File: 133-clone-graph/clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {

    public Node clone(Node node, HashMap<Node, Node> mp)
    {

        if(node == null)
            return null;
        
        if(mp.containsKey(node))
            return mp.get(node);
        
        Node newNode = new Node(node.val);
        mp.put(node, newNode);

        for(Node n : node.neighbors)
        {
            if(!mp.containsKey(n))
            {
                Node node1 = clone(n, mp);
                mp.put(n, node1);
            }
            newNode.neighbors.add(mp.get(n));
        }
        return newNode;
    }
    public Node cloneGraph(Node node) {

        HashMap<Node, Node> mp = new HashMap<>();
        return clone(node, mp);
    }
}

File: 133-clone-graph/clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:

    def clone(self, node: Optional['Node'], map1) -> Optional['Node']:

        if not node:
            return node
        
        if node in map1:
            return map1[node]
        
        map1[node] = Node(node.val)

        for neigh in node.neighbors:
            if neigh in map1:
                map1[node].neighbors.append(map1[neigh])
            else:
                map1[node].neighbors.append(self.clone(neigh, map1))
        
        return map1[node]
            

    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        
        map1 = dict()
        return self.clone(node, map1)
